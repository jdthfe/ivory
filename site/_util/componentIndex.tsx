export default [
    { name: 'CheckUA', type: 'Function', 'zh-CN': '校验UA' },
    { name: 'DeepObjectMerge', type: 'Function', 'zh-CN': '数据扁平化' },
    { name: 'FormatePrice', type: 'Function', 'zh-CN': '格式化价钱' },
    { name: 'GetQuery', type: 'Function', 'zh-CN': '解析URL' },
    { name: 'Throttle', type: 'Function', 'zh-CN': '待输入_中文名' },
];
