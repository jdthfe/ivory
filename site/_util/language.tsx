// en-US must be the first item
export interface LanguageItem {
    code: Language;
    content: string;
}
export const languages: LanguageItem[] = [
    { code: 'en-US', content: 'English' },
    { code: 'zh-CN', content: '中文' },
];

/** 判断语言 */
export function getLanguage(): Language {
    const code = String(localStorage.getItem('language')) as Language;
    if (languages.map((item) => item.code).includes(code)) {
        return code;
    } else {
        return 'en-US';
    }
}
