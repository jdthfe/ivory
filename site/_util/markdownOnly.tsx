export default [
    { name: 'Introduce', type: 'markdownOnly', 'zh-CN': '介绍' },
    { name: 'Customize Theme', type: 'markdownOnly', 'zh-CN': '主题' },
    { name: '_util', type: 'markdownOnly', 'zh-CN': '工具' },
];
