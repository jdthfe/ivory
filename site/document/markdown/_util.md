Common variables in components

```tsx
export const prefix = 'eui';
export const Variable = {
    defaultDuration: 2000,
    transitionTime: 160,
    transitionFade: `${prefix}-fade`,
    transitionZoom: `${prefix}-zoom`,
};

export const removeNode = (el: HTMLDivElement) => {
    if (el.remove) {
        el.remove();
    } else if (el.parentNode) {
        el.parentNode.removeChild(el);
    }
};

export default {
    prefix,
    Variable,
    removeNode,
};
```

Usage

```tsx
import { _util } from '@jdthfe/eui';
// import _util from 'eui/compiled/lib/_util/'; // If you use babel-plugin-import

const { Variable } = _util;
console.log(Variable.defaultDuration); // 2000
```
