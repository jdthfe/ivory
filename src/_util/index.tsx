export const prefix = 'eui';
export const Variable = {
    defaultDuration: 2000,
    transitionTime: 160,
    transitionFade: `${prefix}-fade`,
    transitionZoom: `${prefix}-zoom`,
};

export const removeNode = (el: HTMLDivElement) => {
    if (el.remove) {
        el.remove();
    } else if (el.parentNode) {
        el.parentNode.removeChild(el);
    }
};

export default {
    prefix,
    Variable,
    removeNode,
};
