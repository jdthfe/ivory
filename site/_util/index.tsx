import componentIndexNoType from './componentIndex';
import markdownOnlyNoType from './markdownOnly';
import { languages } from './language';

export interface Docs {
    readme: string;
    'readme.zh-CN'?: string;
    demoSource?: string;
    [key: string]: string | undefined;
}

export const componentIndex: ComponentIndex[] = componentIndexNoType;
export const markdownOnly: ComponentIndex[] = markdownOnlyNoType;

export const sortComponentIndex = (componentIndex: ComponentIndex[]) => {
    const newStructure: { [propName: string]: ComponentIndex[] } = {};
    componentIndex.map((item) => {
        if (newStructure[item.type]) {
            newStructure[item.type].push(item);
        } else {
            newStructure[item.type] = [item];
        }
    });
    return newStructure;
};
export const getInstance = (item: ComponentIndex) => {
    let instance: React.FC;
    try {
        if (item.type === 'markdownOnly') {
            throw 'markdownOnly have not Instance';
        } else {
            instance = require(`../../src/${item.name}/demo`).default;
        }
    } catch (e) {
        throw e;
    }
    return instance;
};
export const getDocs = (item: ComponentIndex) => {
    const docs: Docs = { readme: '' };
    try {
        if (item.type === 'markdownOnly') {
            languages.map((lang) => {
                const readmeLang = `readme${
                        lang.code === 'en-US' ? '' : '.' + lang.code
                    }`,
                    mdLang = `${item.name}${
                        lang.code === 'en-US' ? '' : '.' + lang.code
                    }`;
                docs[readmeLang] =
                    require(`!!raw-loader!../document/markdown/${mdLang}.md`).default;
            });
        } else {
            docs.demoSource =
                require(`!!raw-loader!../../src/${item.name}/demo`).default;
            languages.map((lang) => {
                const readmeLang = `readme${
                    lang.code === 'en-US' ? '' : '.' + lang.code
                }`;
                docs[readmeLang] =
                    require(`!!raw-loader!../../src/${item.name}/demo/${readmeLang}.md`).default;
            });
        }
    } catch (e) {
        throw e;
    }
    return docs;
};
