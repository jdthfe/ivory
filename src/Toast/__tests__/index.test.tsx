import { Toast } from '@src/index';
import React from 'react';
import { render, waitForElement, waitForElementToBeRemoved, fireEvent, cleanup } from 'react-testing-library';

import { prefix } from '../../_util/';

const fadeCls = `.${prefix}-fade-entry-done`;
const prefixCls = `.${prefix}-toast`;

afterEach(cleanup);
describe('Toast', () => {
    const toastTxt = 'toastTxt';
    const duration = 1000;
    const time = 200;

    it('base snapshot about show and close', async () => {
        const { baseElement, getByText } = render(<div />);
        // 1
        // model 不传值自动阻止
        Toast.model();

        const alert1 = Toast.alert();
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}-alert`));
        const success1 = Toast.success();
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}-success`));

        //  排除 svg 文件干扰
        baseElement.querySelector('svg')!.remove();
        expect(baseElement).toMatchSnapshot('show: object = default');
        alert1();
        await waitForElementToBeRemoved(() => baseElement.querySelector(`${prefixCls}-alert`));
        success1();
        await waitForElementToBeRemoved(() => baseElement.querySelector(`${prefixCls}-success`));
        expect(baseElement).toMatchSnapshot('close: object = default');

        // 2
        // duration = 0 Toast 会一直持续，只能手动关闭
        const model = Toast.model({ children: 'model', duration: 0, time });
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}`));

        const success = Toast.success({ children: 'success', duration: 0, time });
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}-success`));
        const alert = Toast.alert({ children: 'alert', duration: 0, time });
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}-alert`));
        Toast.loading();
        await waitForElement(() => baseElement.querySelector(`${fadeCls}${prefixCls}-loading`));
        //  排除 svg 文件干扰
        baseElement.querySelector('svg')!.remove();
        expect(baseElement).toMatchSnapshot('show: object has value');

        model();
        await waitForElementToBeRemoved(() => getByText('model'));
        success();
        await waitForElementToBeRemoved(() => baseElement.querySelector(`${prefixCls}-success`));
        alert();
        await waitForElementToBeRemoved(() => baseElement.querySelector(`${prefixCls}-alert`));
        Toast.closeLoading();
        await waitForElementToBeRemoved(() => baseElement.querySelector(`${prefixCls}-loading`));

        expect(baseElement.querySelectorAll(prefixCls).length).toEqual(0);
    });
    const timeDeviation = 200;

    it('closed successfully', async () => {
        const { getByText } = render(<div />);
        const close = Toast.model({ children: toastTxt, duration });
        const showTime = new Date().getTime();
        await waitForElement(() => getByText(toastTxt));
        close();
        await waitForElementToBeRemoved(() => getByText(toastTxt));
        const realDuration = new Date().getTime() - showTime;
        expect(realDuration <= timeDeviation + time).toEqual(true);
    });

    const CoverClass = 'CoverClass';
    it('cover boolean', async () => {
        const { baseElement, getByText } = render(<div />);
        const mockFn = jest.fn();
        const close = Toast.model({
            children: toastTxt,
            coverProps: { onClick: mockFn, className: CoverClass },
        });
        await waitForElement(() => baseElement.querySelector('.' + CoverClass));
        fireEvent.click(baseElement.querySelector('.' + CoverClass)!);
        expect(mockFn).toBeCalled();
        close();
        await waitForElementToBeRemoved(() => baseElement.querySelector('.' + CoverClass));
        const close2 = Toast.model({
            children: toastTxt,
            noCover: true,
            coverProps: { onClick: mockFn, className: CoverClass },
        });
        await waitForElement(() => getByText(toastTxt));
        expect(baseElement.querySelector('.' + CoverClass)).toEqual(null);
        close2();
    });

    it('loading', async () => {
        const loadingText = 'loading';
        const { getByText, baseElement } = render(<div />);
        Toast.loading({ children: loadingText, duration, time, noCover: true, coverProps: { className: CoverClass } });

        const showTime = new Date().getTime();
        await waitForElement(() => getByText(loadingText));
        baseElement.querySelector('svg')!.remove();

        expect(baseElement.querySelector('.' + CoverClass)).toEqual(null);

        await waitForElementToBeRemoved(() => getByText(loadingText));
        const realDuration = new Date().getTime() - showTime;
        expect(realDuration >= duration && realDuration < duration + time + timeDeviation).toEqual(true);
    });

    it('base', async () => {
        const { baseElement } = render(<Toast visible>{toastTxt}</Toast>);
        expect(baseElement).toMatchSnapshot('show base');
    });
});
