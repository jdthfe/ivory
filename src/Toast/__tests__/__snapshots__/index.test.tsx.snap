// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast base snapshot about show and close: close: object = default 1`] = `
<body>
  
  
  

  <div>
    <div />
  </div>
  <div />
</body>
`;

exports[`Toast base snapshot about show and close: show: object = default 1`] = `
<body>
  
  
  

  <div>
    <div />
  </div>
  <div />
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-toast-alert eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#alert"
        />
      </svg>
    </div>
  </div>
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-toast-success eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#successful"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Toast base snapshot about show and close: show: object has value 1`] = `
<body>
  
  
  

  
  
  

  <div>
    <div />
  </div>
  <div />
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-fade-entry-done eui-fade"
    >
      model
    </div>
  </div>
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-toast-success eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#successful"
        />
      </svg>
      <div
        class="eui-whitespace eui-whitespace-l"
      />
      success
    </div>
  </div>
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-toast-alert eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl"
        fill="#fff"
      >
        <use
          xlink:href="#alert"
        />
      </svg>
      <div
        class="eui-whitespace eui-whitespace-l"
      />
      alert
    </div>
  </div>
  <div>
    <div
      class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
    />
    <div
      class="eui-toast eui-toast-loading eui-fade-entry-done eui-fade"
    >
      <svg
        aria-hidden="true"
        class="eui-icon eui-icon-xl eui-icon-loading"
        fill="#fff"
      >
        <use
          xlink:href="#loading"
        />
      </svg>
    </div>
  </div>
</body>
`;

exports[`Toast base: show base 1`] = `
<body>
  
  
  

  
  
  

  
  
  

  <div />
  <div
    class="eui-cover eui-cover-transparent eui-fade-entry-done eui-fade"
  />
  <div
    class="eui-toast eui-fade-entry-done eui-fade"
  >
    toastTxt
  </div>
</body>
`;
